FSE Navigation Sidebar: Move navigation sidebar in DOM hierarchy

jeyip opened the pull request:
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/master/docs/contributors/repository-management.md#pull-requests. -->

## Description
<!-- Please describe what you have changed or added -->
The focus order for the navigation panel wasn't behaving as expected. Shift+Tab from the back button did not focus the "Toggle Navigation" button, but rather the last element of the header toolbar instead.

Although consistent with the Inserter's behavior, Shaun has identified inserter behavior as a bug ([source](https://github.com/WordPress/gutenberg/pull/25622#issuecomment-699091615)). The goal is to update the placement of the navbar toggle and panel so that focus navigation behaves as expected.

## How has this been tested?
<!-- Please describe in detail how you tested your changes. -->
<!-- Include details of your testing environment, tests ran to see how -->
<!-- your change affects other areas of the code, etc. -->
**Site Editor**
1. Set up the site editor locally
2. Navigate to the site editor
3. Click on the navigation sidebar toggle
4. Smoke test the navigation sidebar
    - Navigation panel closes correctly when toggle is used again
    - Dashboard button works correctly
    - Navigation panel items behave as expected (click on templates and template parts)
    - Styling matches what is in production
    - Responsiveness matches what is in production
    - After clicking on the navigation toggle, user can press the Shift + Tab keys to refocus the navigation toggle

**Post Editor**
1. Navigate to the post editor
2. Smoke test the page layout:
  - Page styling matches what is in production
  - Page layout matches what is in production
  - General responsiveness matches what is in production
  - **Header**
      - Is the header responsive? 
      - Is the layout correct? 
      - Is it fixed to the screen?
  - **Body**
      - Is the body responsive? 
      - Can the block editor scroll?
      - Is the layout correct? 
  - **Footer**
      - Is the header responsive? 
      - Is the layout correct? 
      - Is it fixed to the screen?
  - **Left Sidebar (e.g. block inserter)**
      - Is the layout correct? 
      - Is it fixed to the screen?
  - **Right Sidebar (e.g. block settings)**
      - Is the layout correct? 
      - Is it fixed to the screen?

## Screenshots <!-- if applicable -->
| Navigation Toggle | Navigation Panel |
|---|---|
| <img width="1282" alt="Screen Shot 2020-10-06 at 4 49 05 PM" src="https://user-images.githubusercontent.com/5414230/95271974-217a2600-07f4-11eb-82e7-e7d7416e22ee.png"> | <img width="1280" alt="Screen Shot 2020-10-06 at 4 49 11 PM" src="https://user-images.githubusercontent.com/5414230/95271938-0c9d9280-07f4-11eb-9f77-8ea13a8b199a.png"> |

### Before
![2020-10-07 16 27 36](https://user-images.githubusercontent.com/5414230/95398250-2c978980-08ba-11eb-98c9-67406523e46c.gif)

### After
![2020-10-07 16 25 01](https://user-images.githubusercontent.com/5414230/95398111-d6c2e180-08b9-11eb-9074-b76b1601e2f9.gif)

## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->
Bug fix #25701

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [ ] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR. <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/master/docs/contributors/native-mobile.md -->

jeyip committed:
Set toggle and nav panel as first element in interface skeleton

jeyip committed:
Position sidebar toggle alongside header topbar

jeyip commented:
**Note:**

This PR is still a work in progress. When the navigation panel is open, it doesn't shift sibling content correctly.

jeyip committed:
Add site and content wrappers to interface skeleton

Previously, the navigation panel and navigation toggle lived within skeleton__body and skeleton__header, respectively. In order to address unexpected keyboard focus order between the nav toggle and nav panel, we create a new element inside of interface-skeleton called skeleton__navigation-sidebar.

To account for the layout changes of introducing a new element, we add two wrapper classes: skeleton__site and skeleton__content. Both ensure that the skeleton__header and skeleton__body shrink their children correctly when the navigation sidebar is open.

jeyip committed:
Add and remove padding to ensure document settings title is centered

jeyip committed:
Remove toggle from flow of page when sidebar is closed

The navigation sidebar lives alongside the skeleton header and skeleton body. This is to ensure that the children of the body and header shrink when the navigation sidebar is open.

The problem is that when the navigation sidebar is closed, the skeleton body (i.e. the block editor) no longer spans the entire width of the viewport. The sidebar (and toggle) occupy space at the edge of the screen. To address this, we position the navigation toggle absolutely when the sidebar is closed.

jeyip committed:
Refactor navigation sidebar directory structure

Because navigation panel and navigation toggle live outside the header and body, we place both components in their own navigation sidebar directory

jeyip committed:
Update scss references

jeyip committed:
Update navbar-toggle references

jeyip committed:
Modify z index of footer

jeyip commented on packages/interface/src/components/interface-skeleton/index.js:
```diff
@@ -64,7 +64,7 @@ function InterfaceSkeleton( {
 		>
 			{ /* TODO: Refactor flex layout so that content respects expanded sidebar */ }
-			<div style={ { display: 'flex' } }>
+			<div className="interface-interface-skeleton__layout">
```
**Note:**
Previously, the navigation panel and navigation toggle lived within skeleton__body and skeleton__header, respectively. In order to address unexpected keyboard focus order between the nav toggle and nav panel, we create a new element inside of interface-skeleton called skeleton__navigation-sidebar.

To account for the layout changes of introducing a new element, we add two wrapper classes: skeleton__site and skeleton__content. Both ensure that the skeleton__header and skeleton__body shrink their children correctly when the navigation sidebar is open.

jeyip committed:
Remove unnecessary comment

jeyip commented on packages/interface/src/components/interface-skeleton/index.js:
```diff
@@ -60,55 +63,68 @@ function InterfaceSkeleton( {
-			) }
-			<div className="interface-interface-skeleton__body">
-				{ !! leftSidebar && (
+			<div className="interface-interface-skeleton__layout">
```
Previously, the navigation panel and navigation toggle lived within skeleton__body and skeleton__header, respectively. In order to address unexpected keyboard focus order between the nav toggle and nav panel, we create a new element inside of interface-skeleton called skeleton__navigation-sidebar.

To account for the layout changes of introducing a new element, we add two wrapper classes: skeleton__site and skeleton__content. Both ensure that the skeleton__header and skeleton__body shrink their children correctly when the navigation sidebar is open.

jeyip commented on packages/edit-site/src/components/navigation-sidebar/navigation-toggle/style.scss:
```diff
@@ -6,6 +6,8 @@
 		background: $gray-900;
 		border-radius: 0;
 		display: flex;
+		position: absolute;
```
**Note:**
When the navigation sidebar is closed, the skeleton body (i.e. the block editor) no longer spans the entire width of the viewport. The sidebar (and toggle) occupy space at the left edge of the screen. To address this, we position the navigation toggle absolutely when the sidebar is closed.

jeyip commented on packages/interface/src/components/interface-skeleton/style.scss:
```diff
@@ -130,6 +141,11 @@ html.interface-interface-skeleton__html-container {
+	bottom: 0;
+	width: 100%;
+	background-color: $white;
+	z-index: 90;
```
**Note:**

The sidebar border was overlapping the footer, which is why we set an explicit z-index of 90. I'm not certain where this originates from, but I'll be looking more into it.

david-szabo97 replied:
I just found this file: https://github.com/WordPress/gutenberg/blob/master/packages/base-styles/_z-index.scss  See line 83-84-85

Might be helpful for you and we could add this to the mapping with a comment explaining the value.

Copons replied:
Good point @david-szabo97, and it's a good reminder to use the [`z-index`](https://github.com/WordPress/gutenberg/blob/a8b872408e751b78ad9790ddc1f0446c0577e32e/packages/base-styles/_z-index.scss#L143-L149) mixin (and add the key to the list) when defining new z-indices.

Copons commented on packages/edit-site/src/components/navigation-sidebar/index.js:
```diff
@@ -0,0 +1,19 @@
+				onClick={ onNavigationToggle }
+			/>
+			{ content === 'navigation' && isOpen && <NavigationPanel /> }
+		</>
```
Should we keep the `MainDashboardButton.Slot` removed from `packages/edit-site/src/components/header/index.js`?

vindl replied:
I think it should now wrap `Back to Dashboard` button, instead of the main one that toggles the sidebar. That's what it does in post editor too.

Copons commented:
The behaviour here feels good to me, and the keyboard navigation is exactly as I would expect.

As mentioned elsewhere, I'd even go as far as removing the [auto-focus](https://github.com/WordPress/gutenberg/blob/c1390e0bdd09bd73125317a7f2320ca78726e725/packages/edit-site/src/components/left-sidebar/navigation-panel/index.js#L27-L33) on the first button inside the navigation (Back to Dashboard) on sidebar toggle, since now it's only as far as a simple tab press.
I'd defer this decision to a11y, though.

---

Most obvious change required: replace all the new z-index values with the `z-index` mixin, as mentioned in an inline comment.

---

About the `InterfaceSkeleton` change.

I don't see any obvious regressions, but I'm wondering if there's a way to tidy it up a little bit.
(Note that I haven't tested my propositions, so I'm not certain they would work 🙂)

- Is the `interface-interface-skeleton__layout` wrapper essential?
Couldn't we just use the root `interface-interface-skeleton` as container?
Off the top of my head, the reason for `__layout` is so that the `interface-interface-skeleton__footer` is not affected by the navigation sidebar sliding in.
What if, instead, the sidebar "moved" the footer as well, exactly like it moves header and content?
(Note that this is possibly kinda awkward, considering that status bars are usually full-width unaffected by any other UI element of the window.)

- How do you feel about `interface-interface-skeleton__editor` instead of `interface-interface-skeleton__site`?

- Is "navigation sidebar" a good and descriptive name for the new skeleton element?
I wonder if "navigation" is too semantically specific, whereas I guess we should probably just convey an "interface" meaning (e.g. just "sidebar").
For example, "Drawer" sounds like a good and descriptive name to me: it's established as a UI element (e.g. Google's Material), and it conveys a "sliding" movement, different from the other sidebars, who just simply appear.

jeyip committed:
Remove interface-interface-skeleton__layout

jeyip committed:
Use interface-skeleton as container

jeyip commented:
> Is the interface-interface-skeleton__layout wrapper essential?

Great catch! You were spot on about why I initially implemented `skeleton-layout`, but it turns out that we can remove the layout wrapper while still preserving current behavior. `skeleton-footer` has position `fixed`, which always attaches it to the bottom of the screen 🙂

> How do you feel about interface-interface-skeleton__editor instead of interface-interface-skeleton__site?

`editor` is more semantically specific but it conveys a lot more meaning. I'm on board.

> Is "navigation sidebar" a good and descriptive name for the new skeleton element?
I wonder if "navigation" is too semantically specific, whereas I guess we should probably just convey an "interface" meaning (e.g. just "sidebar"). For example, "Drawer" sounds like a good and descriptive name to me: it's established as a UI element (e.g. Google's Material), and it conveys a "sliding" movement, different from the other sidebars, who just simply appear.

Agreed. I'll give this more thought and circle back tomorrow. For now, I'll use drawer instead of navigation-sidebar.

jeyip committed:
Rename skeleton__site to skeleton_editor

jeyip committed:
Rename skeleton__navigation sidebar to drawer

jeyip committed:
Remove comment

jeyip committed:
Use z-index mixin for interface-skeleton__footer

jeyip committed:
Use z-index mixin for navigation toggle

Copons committed:
Add missing comma in z-index file

Copons commented on packages/edit-site/src/components/editor/index.js:
```diff
@@ -200,6 +201,21 @@ function Editor() {
+															isNavigationOpen
+														}
+														onNavigationToggle={ () =>
+															toggleLeftSidebarContent(
```
(More of a pondering than a request)

Given that now the navigation sidebar is completely separated from the left sidebar, I think they should be handled by different states (e.g. `setLeftSidebar` and `setNavigationSidebar`).

With different states we could even have both sidebar visible at the same time.
I don't know if it's something we want, though, as we might end up with 3 sidebars open at the same time, which feels very cluttered to me.

Either way, we can make navigation and left sidebars "interdependent": when you open one, automatically close the other.

noahtallen replied:
I wanted to put the nav state into redux so I could use it in a random component, so i tried this out here: https://github.com/WordPress/gutenberg/pull/26003

Copons commented on packages/edit-site/src/components/navigation-sidebar/navigation-panel/style.scss:
```diff
@@ -1,7 +1,6 @@
 .edit-site-navigation-panel {
 	animation: edit-site-navigation-panel__slide-in 0.1s linear;
 	height: 100%;
```
(Not caused by this PR, but it's more a mental note to address it at some point)

The panel is 100% height, but in fact it should be slightly shorter: it's pushed down by the header (60px + 1px of border), and it's supposed to end above the footer (25px, only on "> mobile").

I guess we could remove those 86px from the height with `calc()` or as a bottom padding.

Somewhat related: the Navigation component hides both overflows, while it only really needs to hide `overflow-x`.
Hiding `overflow-y` means that if the menu is taller than the screen, it gets cut off.

Copons commented on packages/edit-site/src/components/editor/index.js:
```diff
@@ -200,6 +201,21 @@ function Editor() {
 											<SidebarComplementaryAreaFills />
 											<InterfaceSkeleton
 												labels={ interfaceLabels }
+												drawer={
```
I think we might want to add a custom label for the drawer in `interfaceLabels`.

Like, the interface element is called "Drawer" but it's filled by "Navigation Sidebar" (or at least I think that's the point of `interfaceLabels` 😄).

jeyip replied:
Good catch!

Copons reviewed (commented):
This is looking good to me!

I've left a few comments: all minor stuff and discussion points, possibly to keep in mind for follow ups!

I'd love to have a few more 👀 on the interface changes, as this is the first time I've dabbled with it.

jeyip committed:
Update interface drawer label for navigation sidebar

vindl commented:
This is working as expected for me. The PR now needs a rebase before we can merge it.

shaunandrews commented:
Very cool. Keyboard focus moves as I expected.

--

Maybe not entirely related to this PR, but the transition feels a little wonky. It feels strange as the other canvas and top toolbar aren't being transitioned.

<Img src="https://user-images.githubusercontent.com/191598/95890351-a9f74a00-0d51-11eb-9d1e-a92d28f352a8.gif" width="200"/>

jeyip commented:
> This is working as expected for me. The PR now needs a rebase before we can merge it.

@vindl Apologies for the delay! Rebasing it now.

jeyip committed:
Recenter document actions

jeyip committed:
Create variable for header toolbar min width

jeyip commented:
> Very cool. Keyboard focus moves as I expected.

🎉 

> Maybe not entirely related to this PR, but the transition feels a little wonky. It feels strange as the other canvas and top toolbar aren't being transitioned.

Good observation. I think it may be caused by this PR, but I'm not sure either. I'll take a look.

jeyip commented:
> Maybe not entirely related to this PR

@shaunandrews Yeah I don't think it's this PR in particular. I'm not seeing transitions on the master branch. I agree that the animation feels wonky when everything else jumps. How would we feel about creating a follow-up issue to be prioritized accordingly?

Copons commented:
This looks good to me!
@jeyip I think we should wait for #26003 to merge so that we can use the Redux state for the sidebars.

I'll also create two new issues to tackle the style issues emerged from this PR:

- Review the Nav sidebar animation in relation to the rest of the content.
Currently: on mount, the sidebar starts "outside" the screen, but the rest of the content immediately repositions to make room for the sidebar. I guess everything should move at the same time (sidebar and nav toggle sliding in from the left; content gradually decreasing width).

- See https://github.com/WordPress/gutenberg/pull/25884#discussion_r502370320
  - The Nav _component_ should only hide the `overflow-x`. Hiding the `overflow-y` means that if the content is > the height, its cut off.
  - The Nav _sidebar_ should calculate its height, subtracting the header and footer (except on mobile, where the footer is hidden).

jeyip commented:
> This looks good to me!
@jeyip I think we should wait for #26003 to merge so that we can use the Redux state for the sidebars.

Sounds good! Let me know if you need any help / want me to draft the two issues you mentioned. I'm happy to help 🙂

Copons commented:
> Sounds good! Let me know if you'd like any assistance / want me to draft either or both of the two issues you mentioned. I'm happy to help 🙂

Apologize, I got stuck in various rabbit holes, but I've opened the issues now! #26119 #26120

jeyip committed:
Use navigation sidebar redux actions and selectors in refactored nav sidebar

Navigation sidebar and inserter sidebar state were both moved into the 'core/edit-site' data store. We incorporate any relevant actions and selectors into our source code updates.

noahtallen commented on packages/edit-site/src/components/header/index.js:
```diff
@@ -25,39 +22,38 @@ import SaveButton from '../save-button';
+				__experimentalGetPreviewDeviceType,
+				isFeatureActive,
+				getTemplateId,
+				isNavigationOpened,
```
Looks like we can remove this since we aren't using it any more

jeyip replied:
Good catch

noahtallen commented on packages/edit-site/src/components/header/index.js:
```diff
@@ -25,39 +22,38 @@ import SaveButton from '../save-button';
+					'wp_template',
+					_templateId
+				),
+				isNavigationOpen: isNavigationOpened(),
```
same as above (can be removed)

noahtallen commented on packages/edit-site/src/components/navigation-sidebar/index.js:
```diff
@@ -0,0 +1,26 @@
+const NavigationSidebar = () => {
+	const { isNavigationOpen } = useSelect( ( select ) => {
+		const { isNavigationOpened } = select( 'core/edit-site' );
+		return { isNavigationOpen: isNavigationOpened() };
```
we can simplify this slightly by returning the value directly:


```js
const isNavigationOpen = useSelect( (select) => {
	return select( 'core/edit-site' ).isNavigationOpened();
} );
```

noahtallen commented on packages/edit-site/src/components/navigation-sidebar/navigation-toggle/index.js:
```diff
@@ -1,12 +1,12 @@
 import { __ } from '@wordpress/i18n';
 import { wordpress } from '@wordpress/icons';
 
-function NavigationToggle( { icon, isOpen, onClick } ) {
```
since we no longer pass the `onClick` handler through, should we also get `isOpen` from the state directly here instead of via props? (not sure myself, up for discussion)

jeyip replied:
`NavigationToggle`'s direct parent, `NavigationSidebar` also uses the isOpen prop (which resolves to isNavigationOpened). It felt odd having identical selectors in both the parent and child, but I don't have a strong opinion here.

noahtallen reviewed (commented):
This works well for me in my testing, nice work! I added a few nitpicky comments re. the redux state integration. 

Some thoughts on keyboard navigation:
- should `esc` close the sidebar (and inserter sidebar for that matter)
- In the quick inserter, tab moves between the different sections, and then arrow keys move between blocks. Should we let tab toggle between top bar and nav panel, and then arrow keys to move between actions in the nav panel? Unsure what best practice is here.

jeyip committed:
Remove unnecessary isNavigationOpened selector

jeyip committed:
Return isNavigationOpened directly from selector

Copons reviewed (approved):
Looking good! Thanks @jeyip! ✨

david-szabo97 commented:
Sorry! Those two commits are meant to be committed to a different branch 😬

Copons commented:
> Sorry! Those two commits are meant to be committed to a different branch 😬

No worries! Rebased the PR and dropped the rogue commits.

jeyip commented:
@Copons Thanks for merging this! 🙏

jasmussen commented on packages/base-styles/_variables.scss:
```diff
@@ -56,6 +56,7 @@ $mobile-header-toolbar-height: 44px;
 $mobile-floating-toolbar-height: 44px;
 $mobile-floating-toolbar-margin: 8px;
 $mobile-color-swatch: 48px;
+$header-toolbar-min-width: 335px;
```
@jeyip I know this value is iPhone friendly, but it doesn't appear to work as intended as the mobile experience is otherwise broken. The variable is also used only in this PR and no-where else. I'd recommend not adding this variable, and instead either use `$break-mobile` from _breakpoints.scss so as to avoid fragmenting the responsive dev.

We also have some good mixins that help mobile dev, such as `@include break-small() { ... }` to abstact and keep things in sync.

jeyip replied:
The 335px was meant to ensure a certain width for the toolbar on either side of the header used [here](https://github.com/WordPress/gutenberg/pull/25884/files/4a2937aca4377f3eda568b93d865980e7e72f893#diff-0cda043c56d581dee157884a2abbedb914c3e24d9712cf0317fea8de5a4bf1a9R16-R20) and [here](https://github.com/WordPress/gutenberg/pull/25884/files/4a2937aca4377f3eda568b93d865980e7e72f893#diff-0cda043c56d581dee157884a2abbedb914c3e24d9712cf0317fea8de5a4bf1a9R34-R40). The min width prevented the toolbars from collapsing as the browser viewport was shrunk horizontally. Giving this more thought, it should absolutely be a desktop specific affordance. 

<img width="1344" alt="Screen Shot 2020-10-26 at 3 21 32 PM" src="https://user-images.githubusercontent.com/5414230/97234987-2bfa6080-179f-11eb-842c-178e4b729c5d.png">

I'll open another PR incorporating the break point mixins.

jasmussen replied:
Ah! Thank you for the explanation. I initially pondered where those 335px came from, because it seems a relatively arbitrary number given the base8 grid most of the UI uses. I then thought — maybe it's that somewhat odd iphone resolution which I _remembered_ as 335px (iPhone 6: 750 x 1334, which makes it 375 points horizontally) but as you can see, I was wrong.

With your explanation, though, I think a different approach is needed for min-width. Longer term I think we need to stack the toolbar if there isn't space, but nearer term I think smaller steps can be taken.

> I'm happy to open another PR incorporating the break point mixins.

#26021 does that, and but also hides some buttons, and minimizes labels, as the viewport approaches mobile breakpoints. It's only a first step, and more are needed, but it's a good step. It's currently blocked on the white screen as shown in the intro of the PR, which happens because the left sidebar is always output, regardless of open or not. I could use help with that if you have bandwidth!

> I'm not quite sure I'm tracking here. I think the name of the variable may be misleading, because it sounds like we see the 335px as an iphone friendly breakpoint?

To be specific, we should only be creating new variables in the _variables.scss file if they are meant to be used widely and in multiple places. If it's just for the site editor as is currently the case, I'd just type the content of the variable directly where needed, or just register the variable directly in those files.

I'd look at finding a different approach than the min-width, though (not urgently) it seems a half-step towards solving a problem that needs a more foundational mobile improvement. The work could probably coincide with enforcing of the stacking top toolbar, so we fix this:

<img width="1406" alt="Screenshot 2020-10-27 at 09 55 55" src="https://user-images.githubusercontent.com/1204802/97278966-a926e380-183a-11eb-9cee-a9441f86c12c.png">

I know it's beta! I know there's a lot of work to be done. I'm just:

- excited about the site editor
- of the belief that if we test mobile in every PR, even if it's not superb, we minimize future headaches

jeyip replied:
I added a [commit](https://github.com/WordPress/gutenberg/pull/26021/commits/b4746b5bc04e676134df1a714d51896fff399668) and [a comment](https://github.com/WordPress/gutenberg/pull/26021#issuecomment-717095993) to #26021 that potentially address the issues you bring up here.

jasmussen commented:
This PR appears to consider mobile (specificaly the iPhone based on the 335px variable introduced), but the mobile experience remains broken, so I'm curious how this was tested for mobile. For me, the site editor is just a white screen on mobile, something which #26021 fixes, but even after that fix, the 335px variable appears to not really address any mobile issues. 

I really appreciate all the work you're doing on this, and I know it's super early to optimize for a breakpoint that is likely going to be rarely used for building templates. However this will nevertheless need to be considered at some point, and the sooner we start testing it, the less technical debt we will have to revisit and fix. 

Most of the time it's it's enough to wrap desktop-specific affordances in a `@include break-small() { ... }` clause.

jeyip commented:
> This PR appears to consider mobile (specificaly the iPhone based on the 335px variable introduced), but the mobile experience remains broken, so I'm curious how this was tested for mobile.

Thanks for your thoughts @jasmussen. They're much appreciated.

To clarify, although the PR appears to consider mobile (like the 335px variable), I didn't explicitly take mobile screens into account, which admittedly might be an oversight. I only ensured that the desktop behaved responsively up to the point where the tablet / mobile break-points take over. I've seen the mobile white screen for a while in site editor development, which is part of the reason why I didn't test behavior at smaller screen widths. I should've, however, been more proactive about calling out the unexpected behavior that you're addressing in #26021.

> I really appreciate all the work you're doing on this, and I know it's super early to optimize for a breakpoint that is likely going to be rarely used for building templates. However this will nevertheless need to be considered at some point, and the sooner we start testing it, the less technical debt we will have to revisit and fix.

I think these are very valid points, especially if we plan on supporting mobile site editing in the future. It's tough developing for a mobile experience without an idea of what it will look like, but I agree that we can circumvent a lot of issues with desktop-specific affordances wrapped in breakpoint mixins.

jasmussen commented:
>  It's tough developing for a mobile experience without an idea of what it will look like, but I agree that we can circumvent a lot of issues with desktop-specific affordances wrapped in breakpoint mixins.

Absolutely.

And in absence of full knowledge of that, I'd just test site editor PRs against a viewport that's <600px wide and see how broken it is. As we move forward, we can then raise the bar for what brokenness we're willing to accept.
